Convert zone types field to a bitmask

TeoTwawki opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?
The intention here is to pave the way towards asking the zone itself if it is a "signet zone" and the like. I thought this made more sense here than in the zone misc flags, which are more about what you can do while in a zone. here we want to know if the zone "counts" as these types, so a signet type made sense to me. Note that this is not the same as being a conquest zone. Several things related to signet will function in zones where conquest doesn't happen.
- Changed from enum class to just enum to ease bitwise maths.
- Added zonetypes for signet/sanction/sigil/ionis in the mask.
- Updated all core references to use bitwise checks. values can now be combined.
- Did update existing enum values in zone settings, but did not add new enums to the values in zone_settings yet

## Steps to test these changes
Log in and verify using prints that dungeons still count as dungeons, dynamis still counts as dynamis. There should be no observable changes before and after merging this PR.

TeoTwawki committed:
Convert zone types field to a bitmask
- Changed from enum class to just enum to ease bitwise maths.
- Changed `GetType()` and `getType()` to `GetTypeMask()` and `getTypeMask()`
- Added zonetypes for signet/sanction/sigil/ionis in the mask.
- Updated all core references to use bitwise checks. values can now be combined.
- Did update existing enum values in zone settings, but did not add new enums to the values in zone_settings yet

TODO: template so that zonetype can be enum class and still used as a mask. Would like all enum to be typesafe eventually, but far out of scope for this PR.

zach2good commented:
Suggestion:

Change GetType() to GetTypeMask(), really try to hammer home that its a mask!

TeoTwawki commented:
> Suggestion:
> 
> Change GetType() to GetTypeMask(), really try to hammer home that its a mask!

done, good idea

zach2good commented on src/map/zone.h:
```diff
@@ -415,15 +415,18 @@ enum class CONTINENT_TYPE : uint8
 };
 
-enum class ZONE_TYPE : uint8
+enum ZONETYPE
```
This can still be `enum ZONETYPE : uint16`

zach2good commented on src/map/zone.h:
```diff
@@ -530,7 +533,7 @@ class CZone
 public:
     ZONEID             GetID();
-    ZONE_TYPE          GetType();
+    uint16             GetTypeMask();
```
This can still be `ZONETYPE`, it'll be convertible back into a uint16

zach2good commented on src/map/lua/lua_zone.cpp:
```diff
@@ -453,7 +453,7 @@ void CLuaZone::Register()
     SOL_REGISTER("getName", CLuaZone::getName);
     SOL_REGISTER("getRegionID", CLuaZone::getRegionID);
-    SOL_REGISTER("getType", CLuaZone::getType);
+    SOL_REGISTER("getType", CLuaZone::getTypeMask);
```
This will want to be updated in all the Lua too:  SOL_REGISTER("getType", CLuaZone::getTypeMask);

TeoTwawki replied:
fidn and replace is failing me, brb with multiple corrections after this rebuild

TeoTwawki committed:
Change DEATH_TYPE to enum class

Non classes enums export their contents to the surrounding scope. This causes naming conflicts and presumably is why our predecessors always prefixed everything.

These name conflicts have the side effect of uncovering some very strange scoping in our core - zone scope overlapping with battleentity scope??

TeoTwawki committed:
ZONE_TYPE "UNKNOWN" makes more sense than "NONE"

TeoTwawki committed:
Fix GM commands (missed a few in prior commit)

claywar commented on src/map/utils/zoneutils.cpp:
```diff
@@ -430,9 +430,9 @@ namespace zoneutils
+                            ZONE_TYPE zoneType = PZone->GetTypeMask();
 
-                            if (zoneType != ZONE_TYPE::DUNGEON_INSTANCED)
+                            if (!(zoneType & ZONE_TYPE::INSTANCED))
```
In the future, I think we might want to avoid bit compares in the condition, and perhaps abstract it out to a function.  Out of scope though

TeoTwawki replied:
eventually when I have more time I think we want whats incoming for status effect flag functions: `GetZoneType(Enumhere)` to check single bits of the mask. Just not tonight unless someone else wants to dive in. :)
